Add specs for Email consent opt-in and Unsubscribe page

bcolsson opened the pull request:
Fixes #3109

mathjazz commented:
Not sure if this is ready for review yet, so just adding an implementation note before I forget about it:
We should make the consent text configurable per instance.

mathjazz commented:
As agreed with Bryan, flagging @flodolo and myself for the final review.

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+# Summary
+
+Work required to enable email communication with Pontoon users: opt-in consent user settings, methods for collecting user opt-in, and an unsubscribe page via a single page without logging in.
```
```suggestion
Work required to enable email communication with Pontoon users: opt-in consent user settings, methods for collecting user opt-in, unsubscribe method via a single page that doesn't require login.
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+- Input field: Contact email address
+- Subtitle of input: “If provided, this email address will be used for email communications and will appear under your Profile page instead of the email used for login.”
+- Check box: “Receive updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform.”
+- Consent: “I’m okay with Mozilla handling my personal information as explained in this <a href="https://www.mozilla.org/en-US/privacy/websites/">Privacy Notice</a>."
```
What does "Consent" mean here? Is it a subtitle to the checkbox?

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+- Check box: “Receive updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform.”
+- Consent: “I’m okay with Mozilla handling my personal information as explained in this <a href="https://www.mozilla.org/en-US/privacy/websites/">Privacy Notice</a>."
+
+## Standalone page for email consent shown after existing user log-in
```
```suggestion
## Standalone page for email consent shown after existing user login
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+Upon login (after a user’s previous session expires) an email consent page should be displayed to existing users logging in for the first time after the consent feature is enabled. 
+
+They should be shown two buttons, one to opt-in to emails or one declining to opt-in. This should not be shown again if the user clicks either of the buttons, but should be shown again upon next login if the user closed their tab/browser.
```
```suggestion
They should be shown two buttons, one to opt in to emails or one declining to opt in. This should not be shown again if the user clicks either of the buttons, but should be shown again upon next login if the user closed their tab/browser.
```

Why do we need to wait for the next login, and not display the dialog the next time they load Pontoon?

bcolsson replied:
This could be an implementation detail to check with @mathjazz - but if we show it to them the next time they load Pontoon - does that also mean each time I open a page in a different tab would cause me to be redirected to the consent page? 

I felt doing it on login would be less disruptive to workflows. Since users have to login at least every 2 weeks and the login process forces them out of their workflow already, asking them at that time would encourage more consideration and could potentially improve conversion. (That being said its just a hypothesis and not tested, mostly based on my own personal views.)

flodolo replied:
That's a good point, it can become invasive and they will just dismiss it without much thought.

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+We need to make existing users aware of the new email communication preference.
+
+Upon login (after a user’s previous session expires) an email consent page should be displayed to existing users logging in for the first time after the consent feature is enabled. 
```
Why do we need to wait for a new login? That seems an unnecessary delay.

bcolsson replied:
See my above comment for my reasoning.

mathjazz replied:
I would also do it immediately. It would give results and let us catch any issues faster. It's also what most other sites do.

I'm not too concerned about the page loading in multiple tabs. It could only happen if you open tabs from the URL bar, after the feature ships and before you resolve the page in one of them.

bcolsson replied:
Based on this feedback, I've revised to @mathjazz suggestion.

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+They should be shown two buttons, one to opt-in to emails or one declining to opt-in. This should not be shown again if the user clicks either of the buttons, but should be shown again upon next login if the user closed their tab/browser.
+
+This page also should not be shown if the user has opted-in via their settings.
```
```suggestion
This page also should not be shown if the user has already opted in via their settings.
```

This might be unnecessary if we decide to show the dialog after enabling the feature, without waiting for a new login.

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+Ideally, we would like users to set up their email communication preferences right away upon successful account creation.
+
+After a user successfully creates their Pontoon account, display a standalone page which redirects to the homepage after the user enables or rejects email updates. This page informs the user of successful account creation and request for updating email communication preferences.
```
```suggestion
After a user successfully creates their Pontoon account, display a standalone page which redirects to the homepage after the user enables or rejects email updates. This page informs the user of successful account creation and requests them to update their email communication preferences.
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+After a user successfully creates their Pontoon account, display a standalone page which redirects to the homepage after the user enables or rejects email updates. This page informs the user of successful account creation and request for updating email communication preferences.
+
+They should be shown two buttons, one to opt-in to emails or one declining to opt-in. This should not be shown again if the user clicks either of the buttons, but the standalone page for email consent (above) should be shown again upon next login if the user closed their tab/browser.
```
```suggestion
They should be shown two buttons, one to opt in to emails or one declining to opt in. This should not be shown again if the user clicks either of the buttons, but the standalone page for email consent (above) should be shown again upon next login if the user closed their tab/browser.
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+(Section: content)
+- Header: "Update your email communication preferences"
+- Sub-header: "You can now choose your email communication preferences within Pontoon."
+- Body: "Would you like to receive updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform by email?”
```
…updates **about** the Pontoon platform?

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+As part of new user onboarding, point new users to user preferences and call out the ability to enable email communication preferences.
+
+(Section: content)
+- Header: Stay up to date
```
```suggestion
- Header: "Stay up to date"
```

Maybe we should use backticks/code instead of quotes for text that represents actual UI text (buttons, field names, etc.)?

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+- Header: "Update your email communication preferences"
+- Sub-header: "You can now choose your email communication preferences within Pontoon."
+- Body: "Would you like to receive updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform by email?”
+- Consent: “By enabling email updates, I’m okay with Mozilla handling my personal information as explained in this <a href="https://www.mozilla.org/en-US/privacy/websites/">Privacy Notice</a>."
```
```suggestion
- Consent: “By enabling email updates, I’m okay with Mozilla handling my personal information as explained in this <a href="https://www.mozilla.org/privacy/websites/">Privacy Notice</a>."
```

Fix applies across the document.

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+## Unsubscribe page
+
+Legal regulations (e.g. CAN-SPAM act) have specific requirements to include an unsubscribe link in commercial email messages to manage email preferences without requiring a log-in.
```
```suggestion
Legal regulations (e.g. CAN-SPAM act) have specific requirements to include an unsubscribe link in commercial email messages to manage email preferences without requiring a login.
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+  3) Unsubscribe links in emails will include the user's unique identifier in the URL to correctly identify the user.
+- All options are managed from a single page; I do not need to navigate from the page linked in the email to manage subscriptions.
+- Unsubscribe from all emails with a single button.
+- Subscribe for emails with a single button
```
```suggestion
- Subscribe for emails with a single button.
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,97 @@
+
+Legal regulations (e.g. CAN-SPAM act) have specific requirements to include an unsubscribe link in commercial email messages to manage email preferences without requiring a log-in.
+
+As a user I can:
```
This section is different from the rest of the specs. 

I think we need to call out the various part of the page, including what text and options we're going to display.

bcolsson commented:
@flodolo - I've added a draft of the unsubscribe page content.

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+
+## Email communication user settings
+
+The email opt-in consent option is available under an “Email communications” option in user preferences. “Contact email address” should be moved from Personal information to under this section and an option for non-transactional mass emails opt-ing should be added.
```
```suggestion
The email opt-in consent option is available under the new “Email communications” section in user settings. “Contact email address” is moved from the “Personal information“ section to this section.
```

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+- Section title: `Email communications`
+- Input field: Contact email address
+- Subtitle of input: `If provided, this email address will be used for email communications and will appear under your Profile page instead of the email used for login.`
+- Check box: `Receive updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform.`
```
Please add a note somewhere that all texts need to be configurable per instance, i.e. we should not be hard-coding "Mozilla".

bcolsson replied:
I added it here where you indicated, but let me know if this should be included somewhere else.

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+
+# Feature explanation
+
+## Email communication user settings
```
Are we going to include mockups in the spec?

bcolsson replied:
I created *very* rough mock ups for the opt-in consents for new and existing users.

mathjazz replied:
Thanks, these are very helpful! I think we should streamline the copy on these pages to make the CTA stand out more.

I'm happy to take a stab at that and work on the missing mockups (we should have one at least for the Unsubscribe page).

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+- Sub-header: `You can now choose your email communication preferences within Pontoon.`
+- Body: `Would you like to receive updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform?`
+- Consent: `By enabling email updates, I’m okay with Mozilla handling my personal information as explained in this <a href="https://www.mozilla.org/privacy/websites/">Privacy Notice</a>.`
+- Button: `Enable email updates`
+- Button: `No, thank you`
```
```suggestion
- Primary button: `Enable email updates`
- Secondary button: `No, thank you`
```

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+- Consent: `By enabling email updates, I’m okay with Mozilla handling my personal information as explained in this <a href="https://www.mozilla.org/privacy/websites/">Privacy Notice</a>.`
+- Button: `Enable email updates`
+- Button: `No, thank you`
+- Body: `You can update your email communication preferences at any time in your settings.`
```
Perhaps we should put these into a mockup? "Body", "Consent" and another "Body" don't really tell much.

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+
+(Section: content)
+- Header: `Update your email communication preferences`
+- Sub-header: `You can now choose your email communication preferences within Pontoon.`
```
Are these texts ours or do we need to verify them with legal? I'd encourage users to opt in more actively.

bcolsson replied:
This is great feedback (ideally would like to get this feedback during the earlier review phase). 

I've updated the content for both the new and existing user opt-in page. Flagging @flodolo for visibility.

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+- Unsubscribe from all emails with a single button.
+- Subscribe for emails with a single button.
+- Receive a confirmation that I have successfully unsubscribed/subscribed. I am also instructed how to change my subscriptions later through my user preferences.
+- (In future implementations) Select specific categories of emails to receive / no longer receive. 
```
This should not be part of the spec.

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+  1) Pontoon generates a random unique identifier and assigns it to each user. The algorithm doesn't use any existing user data to generate the ID, which needs to be sufficiently complex to avoid a potential attacker guessing.
+  2) The unsubscribe page can be accessed without login. If a user ID is not provided, or it's incorrect, it will display an error message.
+  3) Unsubscribe links in emails will include the user's unique identifier in the URL to correctly identify the user.
+- All options are managed from a single page; I do not need to navigate from the page linked in the email to manage subscriptions.
```
This doesn't work well with "As a user I can:".

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+ - Header: `Update Your Email Preferences`
+ - Body: `Thank you for contributing to localization at Mozilla. Please use the form below if you would like to stop receiving email updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform.`
+ - Radio button: `Please unsubscribe from all Mozilla localization emails.`
+ - Radio button: `Please keep me sending me emails from Mozilla localization.`
```
```suggestion
 - Radio button: `Please keep sending me emails from Mozilla localization.`
```

mathjazz commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+  2) The unsubscribe page can be accessed without login. If a user ID is not provided, or it's incorrect, it will display an error message.
+  3) Unsubscribe links in emails will include the user's unique identifier in the URL to correctly identify the user.
+- All options are managed from a single page; I do not need to navigate from the page linked in the email to manage subscriptions.
+- Unsubscribe from all emails with a single button.
+- Subscribe for emails with a single button.
+- Receive a confirmation that I have successfully unsubscribed/subscribed. I am also instructed how to change my subscriptions later through my user preferences.
```
I think we can drop all of this, because it is presented with more detail below. Before I saw the actual copy below, I was confused by these bullet points (e.g. why would I want to subscribe on the unsubscribe page, how exactly am I instructed, etc.).

mathjazz reviewed (commented):
Nice work! Left some comments inline.

As a high level note, which should be applied across the entire doc, [note that we use present simple tense across the entire specification](https://github.com/mozilla/pontoon/blob/main/specs/0000-template.md#feature-explanation). So "The link is included" instead of "The link would typically be included".

bcolsson commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,109 @@
+
+Value for users / problems solved:
+- Users can select their preference for email communication
+- Engagement/reach expanded beyond logged-in active Pontoon users
```
I've revised a bit to provide a more accurate statement.

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+
+The email opt-in consent option is available under the new “Email communications” option in user settings. “Contact email address” is moved from the "Personal information" section to this section.
+
+These user email communication preferences are stored and referenced in database to determine if emails should be sent to the user.
```
```suggestion
These user email communication preferences are stored and referenced in the database to determine if emails should be sent to the user.
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+Legal regulations (e.g. CAN-SPAM act) have specific requirements to include an unsubscribe link in commercial email messages to manage email preferences without requiring a login.
+
+As a user I can:
+- Access my unique/personal unsubscribe page through a link without logging in. The link is typically included with non-transactional emails from Pontoon as an “Unsubscribe” link in the footer.
```
```suggestion
- Access my unique/personal unsubscribe page through a link without logging in. The link is typically included with non-transactional emails from Pontoon as an `Unsubscribe` link in the footer.
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+- Access my unique/personal unsubscribe page through a link without logging in. The link is typically included with non-transactional emails from Pontoon as an “Unsubscribe” link in the footer.
+  - Link technical specs:
+  1) Pontoon generates a random unique identifier and assigns it to each user. The algorithm doesn't use any existing user data to generate the ID, which needs to be sufficiently complex to avoid a potential attacker guessing.
+  2) The unsubscribe page can be accessed without login. If a user ID is not provided, or it's incorrect, it will display an error message.
```
```suggestion
  2) The unsubscribe page can be accessed without login. If a user ID is not provided, or it's incorrect, it will display an error message: `User ID not valid. If you copied and pasted the link from an email, make sure to include all characters.`.
```

I think it's worth adding the actual text here in the specs.

bcolsson replied:
Thanks. I think we should avoid calling this a User ID and just say the URL is invalid.

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+ - Title: `Mozilla Localization - Pontoon`
+ - Header: `Update Your Email Preferences`
+ - Body: `Thank you for contributing to localization at Mozilla. Please use the form below if you would like to stop receiving email updates related to localization at Mozilla, projects on Pontoon, and updates to the Pontoon platform.`
+ - Radio button: `Please unsubscribe from all Mozilla localization emails.`
```
```suggestion
 - Radio button: `Please unsubscribe me from all Mozilla localization emails.`
```

?

flodolo replied:
We need to specify which radio option is automatically selected. On the other hand, what about a checkbox, automatically selected, with this text as a label?

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+ - Radio button: `Please keep sending me emails from Mozilla localization.`
+ - Button: `Submit`
+
+ - On submit alert (Unsubscribe): `You've successfully unsubscribed. You can update your preferences anytimes from your <a href="https://pontoon.mozilla.org/settings/">settings</a>.`
```
```suggestion
 - On submit alert (Unsubscribe): `You've successfully unsubscribed. You can update your preferences anytimes from your <a href="https://pontoon.mozilla.org/settings/">settings</a> (make sure that you're logged in to access this page).`
```

flodolo commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+ - Button: `Submit`
+
+ - On submit alert (Unsubscribe): `You've successfully unsubscribed. You can update your preferences anytimes from your <a href="https://pontoon.mozilla.org/settings/">settings</a>.`
+ - On submit alert (Subscribe): `Thanks! You will continue to receive emails from Mozilla localization. You can update your preferences anytimes from your <a href="https://pontoon.mozilla.org/settings/">settings</a>.`
```
If we go the checkbox route, we could have a message like "If you don't want to unsubscribe, or you have clicked the link by mistake, you can close this page and we will make no changes to your settings.`.

bcolsson commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+
+Upon accessing Pontoon for the first time after the feature is enabled, an email opt-in consent page is displayed to existing users.
+
+The user selects their preference by clicking one of two buttons, one to opt in to emails or one declining to opt in. This page is not shown again if the user clicks either of the buttons. The page is shown again if the user navigates away from the page without clicking either of the buttons.
```
I'm updating this as "visit" - @mathjazz seems to indicate that it wouldn't trigger if a link is opened in a new tab, only when opened from the urlbar - please let me know if there's more precise wording that can be used here/

bcolsson commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,130 @@
+As part of new user onboarding, we point new users to user preferences and call out the ability to enable email communication preferences.
+
+(Section: content)
+- Header: `Stay up to date`
```
"Always" here is a bit redundant as "Stay" already implies that, alternates:
Keep up to date
Stay up to date with the latest
Get the latest updates

flodolo replied:
I like "Keep up to date", more dynamic.

bcolsson commented on specs/0117-email-opt-in-unsubscribe.md:
```diff
@@ -0,0 +1,131 @@
+
+Value for users / problems solved:
+- Users can select their preference for email communication
+- Engagement/reach expanded beyond logged-in active Pontoon users that respects user preferences and complies with legal requirements
```
I reflected both comments since the points they raise are both true.

mathjazz commented:
Depending on the opt-in rate we might not be able to reach as many users as we'd like. How do we mitigate this risk?

Should we — after a period of time — also show the opt-in form to users that have declined to opt-in in the past?

bcolsson commented:
> Depending on the opt-in rate we might not be able to reach as many users as we'd like. How do we mitigate this risk?
> 
> Should we — after a period of time — also show the opt-in form to users that have declined to opt-in in the past?

I'm not sure if we'd want to do this too often to respect the user's choice, otherwise we risk looking like a [dark pattern (nagging)](https://dl.acm.org/doi/pdf/10.1145/3173574.3174108). 

In my research on opt-ins, I've read that it's considered best practice to try and regain permission from users who have not interacted with your emails after 6 months. Since we wouldn't have the ability to track that, we could instead show all users at 6 month intervals a screen to confirm their email communication preferences. What do you think about that?

bcolsson commented:
> > Depending on the opt-in rate we might not be able to reach as many users as we'd like. How do we mitigate this risk?
> > Should we — after a period of time — also show the opt-in form to users that have declined to opt-in in the past?
> 
> I'm not sure if we'd want to do this too often to respect the user's choice, otherwise we risk looking like a [dark pattern (nagging)](https://dl.acm.org/doi/pdf/10.1145/3173574.3174108).
> 
> In my research on opt-ins, I've read that it's considered best practice to try and regain permission from users who have not interacted with your emails after 6 months. Since we wouldn't have the ability to track that, we could instead show all users at 6 month intervals a screen to confirm their email communication preferences. What do you think about that?

Following-up off-line with @flodolo, I think in this case we should respect the user's choice at time of setting the preference. To promote additional opt-in and mitigate against this risk we should use our channels (socials, events, etc.) to remind people to sign-up / change their preferences.

mathjazz commented:
Here's my proposal for the designs of the standalone opt-in consent pages. I tried to keep it simple and concise, as that's a general good practice to increase conversion rate. It's also consistent with most of the email subscription forms I've found online, including the newsletter subscription forms on Mozilla websites:
https://blog.mozilla.org/#newsletter-content
https://www.mozilla.org/

All the designs are the same, just the title is different.

I wonder if "By enabling email updates, I'm okay..." should in fact be an off-by-default checkbox. Legal will know.

![1](https://github.com/mozilla/pontoon/assets/626716/93f8a07f-d15c-4496-962d-68a159e8334a)
![2](https://github.com/mozilla/pontoon/assets/626716/0e489a2b-63ca-4978-942b-f35ddf42b6bb)
![3](https://github.com/mozilla/pontoon/assets/626716/f4dd4f82-9014-4233-a791-bf61742505c4)
![4](https://github.com/mozilla/pontoon/assets/626716/62e65a84-c061-4969-925b-dbda24965324)

mathjazz commented:
> I'm not sure if we'd want to do this too often to respect the user's choice, otherwise we risk looking like a [dark pattern (nagging)](https://dl.acm.org/doi/pdf/10.1145/3173574.3174108).
> 
> In my research on opt-ins, I've read that it's considered best practice to try and regain permission from users who have not interacted with your emails after 6 months. Since we wouldn't have the ability to track that, we could instead show all users at 6 month intervals a screen to confirm their email communication preferences. What do you think about that?

Absolutely - we shouldn't do this more often than once or twice per year. We'll need to track when was the consent given or not for each user, so we should add that to the spec.

External channels are typically much less effective in steering users to take an action within the platform, so I'd consider that as a bonus.

bcolsson commented:
@mathjazz - I've updated copy and mockups with suggestions from copy editing. I've also unified the consent ask page to be the same for existing and new users. Unless you have objections, I'm going to present the spec we have here to legal.

I will also be asking their opinion on whether or not asking for consent again (regarding this [open question](https://github.com/mozilla/pontoon/pull/3110#issuecomment-2010570218)).

mathjazz commented:
No objections, looks great!

bcolsson commented:
I've updated the spec with all of the legal feedback.

Regarding the one outstanding question of whether we want to show a reminder to users about opt-ins: there weren't any legal concerns if this happened in-app. (Sending an email about this to users would not be okay however.) So this option remains open to us. However, I haven't added this to this spec document to limit the scope to what we need to start gathering permissions now. 

I have added the requirement that we note that we record the user's choice of opt-in or not, and the date and time they made the choice. This way we can add it in a future update.

mathjazz reviewed (approved):
LGTM
